src/window/update_loop.rs

```diff
@@ -0,0 +1,117 @@
+use std::time::{Duration, Instant};
+
+use winit::{
+    event::{Event, WindowEvent},
+    event_loop::ControlFlow,
+};
+
+#[cfg(target_os = "macos")]
+use super::draw_background;
+use super::{WindowSettings, WinitWindowWrapper};
+use crate::{
+    profiling::{tracy_create_gpu_context, tracy_zone},
+    running_tracker::*,
+    settings::{save_window_size, SETTINGS},
+};
+
+enum FocusedState {
+    Focused,
+    UnfocusedNotDrawn,
+    Unfocused,
+}
+
+pub struct UpdateLoop {
+    previous_frame_start: Instant,
+    focused: FocusedState,
+}
+
+impl UpdateLoop {
+    pub fn new() -> Self {
+        tracy_create_gpu_context("main_render_context");
+
+        let previous_frame_start = Instant::now();
+        let focused = FocusedState::Focused;
+
+        Self {
+            previous_frame_start,
+            focused,
+        }
+    }
+
+    fn get_event_deadline(&self) -> Instant {
+        let refresh_rate = match self.focused {
+            FocusedState::Focused | FocusedState::UnfocusedNotDrawn => {
+                SETTINGS.get::<WindowSettings>().refresh_rate as f32
+            }
+            FocusedState::Unfocused => SETTINGS.get::<WindowSettings>().refresh_rate_idle as f32,
+        }
+        .max(1.0);
+
+        let expected_frame_duration = Duration::from_secs_f32(1.0 / refresh_rate);
+        self.previous_frame_start + expected_frame_duration
+    }
+
+    pub fn step(
+        &mut self,
+        window_wrapper: &mut WinitWindowWrapper,
+        event: Event<()>,
+    ) -> ControlFlow {
+        tracy_zone!("render loop", 0);
+
+        // Window focus changed
+        if let Event::WindowEvent {
+            event: WindowEvent::Focused(focused_event),
+            ..
+        } = event
+        {
+            self.focused = if focused_event {
+                FocusedState::Focused
+            } else {
+                FocusedState::UnfocusedNotDrawn
+            };
+        }
+
+        let deadline = self.get_event_deadline();
+
+        if !RUNNING_TRACKER.is_running() {
+            let window = window_wrapper.windowed_context.window();
+            save_window_size(
+                window.is_maximized(),
+                window.inner_size(),
+                window.outer_position().ok(),
+            );
+
+            std::process::exit(RUNNING_TRACKER.exit_code());
```

Comment:
Not sure if it's best to handle the exit directly here. Possibly it could be worth bubbling it up instead, returning [ControlFlow::ExitWithCode](https://docs.rs/winit/0.29.0-beta.0/winit/event_loop/enum.ControlFlow.html#variant.ExitWithCode) instead.

```suggestion
            return ControlFlow::ExitWithCode(RUNNING_TRACKER.exit_code());
```